Monitord

muhamadazmy opened the pull request.

muhamadazmy committed:
Add basic monitoring

muhamadazmy committed:
Add timestamp to disk io

muhamadazmy committed:
Preparation for UI

muhamadazmy committed:
Rename file to system

muhamadazmy committed:
Add monitor to system version

muhamadazmy committed:
Basic app for ui

Kernel writes over the UI for some reason

muhamadazmy committed:
Show running mode

muhamadazmy committed:
Have a clean UI

muhamadazmy committed:
Graph CPU usage

muhamadazmy committed:
WIP: memory graph

muhamadazmy committed:
Add used and free memory graph

muhamadazmy committed:
Ignore `monitord` spell checking

muhamadazmy committed:
Monitor changes of IP addresses on the zos device

muhamadazmy committed:
Building ui for network info

muhamadazmy committed:
Add network screen, also add uptime monitor

zaibon commented:
I quickly went over the code and I think we could avoid having a new monitord daemon. 
I think each daemon is already responsible for a part of the UI and the data should come from them.

Everythigng related with system stats (cpu, memory) -> capacityd
Everything related with host (node ID, farm, version) -> identityd
Everything related with network -> networkd
Everything related with disks -> storaged

muhamadazmy committed:
Merge both capacityd and monitord in the same binary

This is only to redue mem footprint, since they
both uses the same libraries

muhamadazmy committed:
Move version monitor to identityd

muhamadazmy committed:
move IP monitoring to networkd

muhamadazmy committed:
Generic ns monitor mechanism

muhamadazmy committed:
Add pool/disk stats

Note: for some reason counters do not show up

muhamadazmy committed:
Add missing interface method

muhamadazmy committed:
WIP: provision monitoring

BUG: the counter can go below zero when we receive expired reservations
there is no way to check if this reservation is active and need to be
deleted, or was never provisioned

muhamadazmy commented on pkg/provision/engine.go:
```diff
@@ -135,6 +167,8 @@ func (e *defaultEngine) decommission(ctx context.Context, r *Reservation) error
 		return errors.Wrap(err, "decommissioning of reservation failed")
 	}
 
+	e.stats[r.Type].Decrement()
```
This can/will go below zero (minus value) when we process expired, or to-delete reservations that were never provisioned on the node. There is no way (I know of) to check if the reservation to delete is active, or was never deployed.

zaibon replied:
Well, at line 172, we remove the reservation from the reservation cache.
I think this cache is the most accurate view on what is actually deployed on the node or not. Maybe we could use it to return the number of different workload deployed at any point in time.
This kind of information is something I would also like to send to BCDB at some point.

muhamadazmy committed:
Move counters to the store

muhamadazmy committed:
Text based repr of cpu and memory

muhamadazmy committed:
Finish dropping all graphs

Next, we need to add stats for disks

muhamadazmy committed:
fix tests

muhamadazmy committed:
Add all widgets

muhamadazmy committed:
render max every 1 sec

muhamadazmy committed:
ange layout a little bit

muhamadazmy committed:
Merge branch 'master' into monitord

zaibon commented on etc/zinit/networkd.yaml:
```diff
@@ -1,3 +1,4 @@
 exec: networkd -broker unix:///var/run/redis.sock -root /var/cache/modules/networkd
+test: ip link show zos
```
Internet is the service responsible to create zos,not networkd. So this test will always pass right away and doesn't really bring anything.

muhamadazmy replied:
Yeah, actually i totally can take this out, i needed it before the refactoring. This just grantee that the zos bridge is created for monitoring, but now network monitoring is done by networkd itself. so I will remove that.

zaibon commented on pkg/capacity/capacity.go:
```diff
@@ -17,11 +17,11 @@ type Capacity struct {
 // ResourceOracle is the structure responsible for capacity tracking
 type ResourceOracle struct {
-	storage pkg.StorageModule
+	storage *stubs.StorageModuleStub
```
Why do you replace the interface with the stub type ?

muhamadazmy replied:
because zbus events functions unfortunately have different signature between both the module interface, and the generated stubs, unlike the other methods where they generate the same exact interface.

For example
```
// a module that has this method
func (m *Module) Stream(ctx context.Context) <- chan Value{
    
}

// will generate a stub like
func (m *Stub) Stream(ctx context.Context) (<-chan Value, error) {
}
```

Hence it's not possible to use the interface here.

zaibon commented on pkg/monitord/host.go:
```diff
@@ -0,0 +1,124 @@
+	if duration == 0 {
+		duration = 2 * time.Second
+	}
+	watcher, err := fsnotify.NewWatcher()
```
Do you still needs this watcher here ? Since the version is not pushed from identityd itself when a new version is available.

muhamadazmy replied:
yup, this need to go idd.

zaibon commented on pkg/network/namespace/monitor.go:
```diff
@@ -0,0 +1,25 @@
+	ch := make(chan netlink.AddrUpdate)
+	if err := netlink.AddrSubscribeAt(ns, ch, ctx.Done()); err != nil {
+		close(ch)
+		ns.Close()
```
are we sure this will always be called ? otherwise we could leak a file descriptor here.

muhamadazmy replied:
I checked the AddrSubscribe code, and they close the namespace once the ctx channel is done. Hence it's safe to not close it as long as the subscribe has been successful.

zaibon commented on pkg/provision/engine.go:
```diff
@@ -40,6 +42,11 @@ type defaultEngine struct {
 // one reservation at a time. On error, the engine will log the error. and
 // continue to next reservation.
 func New(source ReservationSource, rw ReservationReadWriter, fb Feedbacker) Engine {
+	stats := make(map[ReservationType]*Counter)
```
This looks like dead code, no ?

muhamadazmy replied:
Good catch :)

zaibon commented on qemu/overlay/bin/monitord:
```diff
@@ -0,0 +1 @@
+../../../bin/monitord
```
this doesn't exist anymore

muhamadazmy replied:
Again, very good catch, man you are on fire :fire:

zaibon commented on qemu/overlay/etc/promtail/promtail.yaml:
```diff
@@ -1 +0,0 @@
-../../../../etc/promtail/promtail.yaml
```
why is this gone ?

muhamadazmy replied:
This is installed from the binaries, it's not part of zos core anymore

zaibon commented on qemu/overlay/etc/zinit/monitord.yaml:
```diff
@@ -0,0 +1 @@
+../../../../etc/zinit/monitord.yaml
```
this doesn't exist anymore

muhamadazmy replied:
removed

muhamadazmy committed:
Apply review comments

Commit:
Merge branch 'master' into monitord
